core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,218 @@
+                                  pollIntervalJitter: FiniteDuration,
+                                  workspaceService: UserInfo => WorkspaceService,
+                                  googleServicesDAO: GoogleServicesDAO,
+                                  samDAO: SamDAO,
+                                  googleStorage: GoogleStorageService[IO],
+                                  pubSubDao: GooglePubSubDAO,
+                                  pubSubSubscriptionName: String,
+                                  avroUpsertBucketName: String)
+  extends Actor
+    with LazyLogging
+    with FutureSupport {
+  import AvroUpsertMonitor._
+  import context._
+
+  case class AvroUpsertJson(namespace: String, name: String, userSubjectId: String, userEmail: String, payload: Array[EntityUpdateDefinition])
+
+  implicit val avroUpsertJsonFormat = jsonFormat5(AvroUpsertJson)
+
+  self ! StartMonitorPass
+
+  // fail safe in case this actor is idle too long but not too fast (1 second lower limit)
+  setReceiveTimeout(max((pollInterval + pollIntervalJitter) * 10, 1 second))
+
+  private def max(durations: FiniteDuration*): FiniteDuration = {
+    implicit val finiteDurationIsOrdered = scala.concurrent.duration.FiniteDuration.FiniteDurationIsOrdered
+    durations.max
+  }
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      logger.info(s"received avro upsert message: $message")
+      val (jobId, file) = parseMessage(message)
+      initAvroUpsert(jobId, file).map(response => (response, message.ackId)) pipeTo self
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime.asInstanceOf[FiniteDuration], self, StartMonitorPass)
+
+    case ((), ackId: String) =>
+      acknowledgeMessage(ackId).map(_ => StartMonitorPass) pipeTo self
+
+    case Status.Failure(t) => throw t
+
+    case ReceiveTimeout =>
+      throw new WorkbenchException("AvroUpsertMonitorActor has received no messages for too long")
+
+    case x =>
+      logger.info(s"unhandled $x")
+//      self ! StartMonitorPass TODO: should this restart the monitor if we've got an unhandled message? I think so.
+  }
+
+  private def initAvroUpsert(jobId: String, file: String): Future[Unit] = {
+    for {
+      avroJson <- readUpsertObject(s"$jobId/$file")
+      petSAJson <- samDAO.getPetServiceAccountKeyForUser(avroJson.namespace, RawlsUserEmail(avroJson.userEmail))
+      petUserInfo <- googleServicesDAO.getUserInfoUsingJson(petSAJson)
+      _ <- workspaceService.apply(petUserInfo).batchUpdateEntities(WorkspaceName(avroJson.namespace, avroJson.name), avroJson.payload.toSeq, true)
+    } yield ()
+  }
```

Comment:
1) for debugging purposes, how about adding a log statement here with the jobid, userEmail, and avroJson.payload.toSeq.size? I'd also love to log the result of `batchUpdateEntities`.

2) since this sends `batchUpdateEntities` to `_` ... does it wait for the Future to complete or not? If not, it will ack the pubsub message and start a new monitor pass while this upsert is potentially still running. If it does wait for the Future, we could hit ReceiveTimeouts on the actor. I'm thinking it's okay to start a new monitor pass before this import completes .... just wanted to raise the issue to make sure it's all intentional